File: Micro-Services/pong_project/pong_project/static/js/pong/game.js
import * as g from './global.js';
import * as physics from './physics.js';

export class Game {
	constructor() {
		this.t = 0.0;
		this.status = g.STATUS_ACTIVE;
		this.scores = [0, 0];
		this.players_ready = [0, 0];
		this.who_serves = g.ID_PLAYER1;
		this.particle_pool = new physics.ParticlePool(16);
		this.ball = new physics.Rectangle((g.BOARD_WIDTH - g.BALL_SIDE) / 2, (g.BOARD_HEIGHT - g.BALL_SIDE) / 2, g.BALL_SIDE, g.BALL_SIDE, 0, g.BALL_SPEED_MIN);
		this.player1 = new physics.Rectangle((g.BOARD_WIDTH - g.PADDLE_WIDTH) / 2, g.BOARD_HEIGHT - (3 * g.BOARD_WALL), g.PADDLE_WIDTH, g.BALL_SIDE, 0, 0);
		this.player2 = new physics.Rectangle((g.BOARD_WIDTH - g.PADDLE_WIDTH) / 2, 2 * g.BOARD_WALL, g.PADDLE_WIDTH, g.BALL_SIDE, 0, 0);
		this.collision_happened = false;
		this.victory_happened = false;
		this.score_happened = false;
		this.reset_ball();
	}

	reset() {
		this.t = 0.0;
		this.status = g.STATUS_ACTIVE;
		this.players_ready = [0, 0];
		this.scores = [0, 0];
		this.who_serves = g.ID_PLAYER1;
		this.particle_pool = new physics.ParticlePool(16);
		this.ball = new physics.Rectangle((g.BOARD_WIDTH - g.BALL_SIDE) / 2, (g.BOARD_HEIGHT - g.BALL_SIDE) / 2, g.BALL_SIDE, g.BALL_SIDE, 0, g.BALL_SPEED_MIN);
		this.player1 = new physics.Rectangle((g.BOARD_WIDTH - g.PADDLE_WIDTH) / 2, g.BOARD_HEIGHT - (3 * g.BOARD_WALL), g.PADDLE_WIDTH, g.BALL_SIDE, 0, 0);
		this.player2 = new physics.Rectangle((g.BOARD_WIDTH - g.PADDLE_WIDTH) / 2, 2 * g.BOARD_WALL, g.PADDLE_WIDTH, g.BALL_SIDE, 0, 0);
		this.collision_happened = false;
		this.victory_happened = false;
		this.score_happened = false;
		this.reset_ball();
	}

	reset_ball() {
		this.ball.position.x = (g.BOARD_WIDTH - this.ball.size.x) / 2;
		this.ball.position.y = (g.BOARD_HEIGHT - this.ball.size.y) / 2;

		const direction = this.who_serves === g.ID_PLAYER1 ? 1 : -1;
		const angle = Math.random() * (g.BALL_MAX_ANGLE * 2) - g.BALL_MAX_ANGLE;

		this.ball.velocity.x = Math.sin(angle) * g.BALL_SPEED_MIN * direction;
		this.ball.velocity.y = Math.cos(angle) * g.BALL_SPEED_MIN * direction;
	}

	/* Update the angle of the ball based on where it hits the paddle */
	update_ball_velocity(paddle, normal) {
		const expanded =
			new physics.Rectangle(
				paddle.position.x - this.ball.size.x / 2,
				paddle.position.y - this.ball.size.y / 2,
				paddle.size.x + this.ball.size.x,
				paddle.size.y + this.ball.size.y,
				0,
				0
			);
		const ball_center =
			new physics.Vector(
				this.ball.position.x + this.ball.size.x / 2,
				this.ball.position.y + this.ball.size.y / 2
			);
		const paddle_center =
			new physics.Vector(
				expanded.position.x + expanded.size.x / 2,
				expanded.position.y + expanded.size.y / 2
			);

		if (normal.x != 0) {
			let c = ((ball_center.y - paddle_center.y) / (expanded.size.y / 2)) * g.BALL_MAX_ANGLE;
			this.ball.velocity.x = normal.x * Math.cos(c) * g.BALL_SPEED_MAX;
			this.ball.velocity.y = Math.sin(c) * g.BALL_SPEED_MAX;
		} else if (normal.y != 0) {
			let c = ((ball_center.x - paddle_center.x) / (expanded.size.x / 2)) * g.BALL_MAX_ANGLE;
			this.ball.velocity.x = Math.sin(c) * g.BALL_SPEED_MAX;
			this.ball.velocity.y = normal.y * Math.cos(c) * g.BALL_SPEED_MAX;
		}
	}

	update_paddle_position(paddle, dt) {
		paddle.position.x += paddle.velocity.x * dt;

		if (paddle.position.x < g.BOARD_CORRIDOR) {
			paddle.position.x = g.BOARD_CORRIDOR;
		} else if (paddle.position.x + paddle.size.x > g.BOARD_WIDTH - g.BOARD_CORRIDOR) {
			paddle.position.x = g.BOARD_WIDTH - g.BOARD_CORRIDOR - paddle.size.x;
		}
	}

	update_ball_position(dt) {
		let player = null;
		let collision = null;

		const c1 = physics.aabb_continuous_detection(this.ball, this.player1, dt);
		const c2 = physics.aabb_continuous_detection(this.ball, this.player2, dt);

		if (c1.time > 0) {
			player = this.player1;
			collision = c1;
		} else if (c2.time > 0) {
			player = this.player2;
			collision = c2;
		}

		if (collision != null && player != null && collision.time > 0 && collision.time <= 1.0) {
			this.collision_happened = true;
			let v = physics.aabb_continuous_resolve(this.ball, collision);
			this.ball.position.x += v.x * dt;
			this.ball.position.y += v.y * dt;
			this.update_ball_velocity(player, collision.normal);
		} else {
			this.ball.position.x += this.ball.velocity.x * dt;
			this.ball.position.y += this.ball.velocity.y * dt;
		}

		if (this.ball.position.x <= g.BOARD_WALL || this.ball.position.x + this.ball.size.x >= g.BOARD_WIDTH - g.BOARD_WALL) {
			/* Left and right walls */
			this.ball.position.x = this.ball.position.x <= g.BOARD_WALL ? g.BOARD_WALL : g.BOARD_WIDTH - this.ball.size.x - g.BOARD_WALL;
			this.ball.velocity.x *= -1;
			this.collision_happened = true;
		} else if (this.ball.position.y <= g.BOARD_WALL || this.ball.position.y + this.ball.size.y >= g.BOARD_HEIGHT - g.BOARD_WALL) {
			/* Top and bottoms walls */
			if (this.ball.position.y <= g.BOARD_WALL)
				this.scores[g.ID_PLAYER1] += 1;
			else
				this.scores[g.ID_PLAYER2] += 1;

			this.particle_pool.reset(this.ball.position.x + this.ball.size.x / 2, this.ball.position.y + this.ball.size.y / 2);
			this.who_serves = this.who_serves === g.ID_PLAYER1 ? g.ID_PLAYER2 : g.ID_PLAYER1;
			this.score_happened = true;
			this.reset_ball();

			if (this.scores[g.ID_PLAYER1] >= g.POINTS_TO_WIN || this.scores[g.ID_PLAYER2] >= g.POINTS_TO_WIN) {
				this.victory_happened = true;
				this.ball.velocity.y = 0;
				this.ball.velocity.x = 0;
			}
		}
	}

	update(dt) {
		this.t += dt;

		if (this.t < 1 || (this.status != g.STATUS_ACTIVE && this.status != g.STATUS_ENDED))
			return;

		this.update_paddle_position(this.player1, dt);
		this.update_paddle_position(this.player2, dt);

		if (this.particle_pool.get_n_actives() > 0) {
			this.particle_pool.update(dt);
			return;
		}

		this.update_ball_position(dt);
	}
}

File: Micro-Services/game_app/game_app/pong/game.py
import math
import random

import game_app.pong.constants as g
import game_app.pong.physics as physics

class Game:
    def __init__(self) -> None:
        self.t = 0.0
        self.status = g.STATUS_WAITING
        self.scores = [0,0]
        self.players_ready = [0, 0]
        self.who_serves = g.ID_PLAYER1
        self.particle_pool = physics.ParticlePool(16)
        self.ball = physics.Rectangle((g.BOARD_WIDTH - g.BALL_SIDE) / 2, (g.BOARD_HEIGHT - g.BALL_SIDE) / 2, g.BALL_SIDE, g.BALL_SIDE, 0, g.BALL_SPEED_MIN,)
        self.player1 = physics.Rectangle((g.BOARD_WIDTH - g.PADDLE_WIDTH) / 2, g.BOARD_HEIGHT - (3 * g.BOARD_WALL), g.PADDLE_WIDTH, g.BALL_SIDE, 0, 0,)
        self.player2 = physics.Rectangle((g.BOARD_WIDTH - g.PADDLE_WIDTH) / 2, 2 * g.BOARD_WALL, g.PADDLE_WIDTH, g.BALL_SIDE, 0, 0)
        self.collision_happened = False
        self.victory_happened = False
        self.score_happened = False
        self.reset_ball()

    def reset(self):
        self.t = 0.0
        self.status = g.STATUS_WAITING
        self.scores = [0,0]
        self.players_ready = [0, 0]
        self.who_serves = g.ID_PLAYER1
        self.particle_pool = physics.ParticlePool(16)
        self.ball = physics.Rectangle((g.BOARD_WIDTH - g.BALL_SIDE) / 2, (g.BOARD_HEIGHT - g.BALL_SIDE) / 2, g.BALL_SIDE, g.BALL_SIDE, 0, g.BALL_SPEED_MIN,)
        self.player1 = physics.Rectangle((g.BOARD_WIDTH - g.PADDLE_WIDTH) / 2, g.BOARD_HEIGHT - (3 * g.BOARD_WALL), g.PADDLE_WIDTH, g.BALL_SIDE, 0, 0,)
        self.player2 = physics.Rectangle((g.BOARD_WIDTH - g.PADDLE_WIDTH) / 2, 2 * g.BOARD_WALL, g.PADDLE_WIDTH, g.BALL_SIDE, 0, 0)
        self.collision_happened = False
        self.victory_happened = False
        self.score_happened = False
        self.reset_ball()

    def reset_ball(self):
        self.ball.position.x = (g.BOARD_WIDTH - self.ball.size.x) / 2
        self.ball.position.y = (g.BOARD_HEIGHT - self.ball.size.y) / 2

        direction = 1 if self.who_serves == g.ID_PLAYER1 else -1
        angle = random.random() * (g.BALL_MAX_ANGLE * 2) - g.BALL_MAX_ANGLE

        self.ball.velocity.x = math.sin(angle) * g.BALL_SPEED_MIN * direction
        self.ball.velocity.y = math.cos(angle) * g.BALL_SPEED_MIN * direction

    # Update the angle of the ball based on where it hits the paddle
    def update_ball_velocity(self, paddle, normal):
        expanded = physics.Rectangle(
            paddle.position.x - self.ball.size.x / 2,
            paddle.position.y - self.ball.size.y / 2,
            paddle.size.x + self.ball.size.x,
            paddle.size.y + self.ball.size.y,
            0,
            0,
        )
        ball_center = physics.Vector(self.ball.position.x + self.ball.size.x / 2, self.ball.position.y + self.ball.size.y / 2)
        paddle_center = physics.Vector(expanded.position.x + expanded.size.x / 2, expanded.position.y + expanded.size.y / 2)

        if normal.x != 0:
            c = ((ball_center.y - paddle_center.y) / (expanded.size.y / 2)) * g.BALL_MAX_ANGLE
            self.ball.velocity.x = normal.x * math.cos(c) * g.BALL_SPEED_MAX
            self.ball.velocity.y = math.sin(c) * g.BALL_SPEED_MAX
        elif normal.y != 0:
            c = ((ball_center.x - paddle_center.x) / (expanded.size.x / 2)) * g.BALL_MAX_ANGLE
            self.ball.velocity.x = math.sin(c) * g.BALL_SPEED_MAX
            self.ball.velocity.y = normal.y * math.cos(c) * g.BALL_SPEED_MAX

    def update_paddle_position(self, paddle, dt):
        paddle.position.x += paddle.velocity.x * dt

        if paddle.position.x < g.BOARD_CORRIDOR:
            paddle.position.x = g.BOARD_CORRIDOR
        elif paddle.position.x + paddle.size.x > g.BOARD_WIDTH - g.BOARD_CORRIDOR:
            paddle.position.x = g.BOARD_WIDTH - g.BOARD_CORRIDOR - paddle.size.x


    def update_ball_position(self, dt):
        player = None
        collision = None

        c1 = physics.aabb_continuous_detection(self.ball, self.player1, dt)
        c2 = physics.aabb_continuous_detection(self.ball, self.player2, dt)

        if c1.time > 0:
            player = self.player1
            collision = c1
        elif c2.time > 0:
            player = self.player2
            collision = c2

        if collision is not None and player is not None and collision.time > 0 and collision.time <= 1.0:
            self.collision_happened = True
            v = physics.aabb_continuous_resolve(self.ball, collision)
            self.ball.position.x += v.x * dt
            self.ball.position.y += v.y * dt
            self.update_ball_velocity(player, collision.normal)
        else:
            self.ball.position.x += self.ball.velocity.x * dt
            self.ball.position.y += self.ball.velocity.y * dt

        if self.ball.position.x <= g.BOARD_WALL or self.ball.position.x + self.ball.size.x >= g.BOARD_WIDTH - g.BOARD_WALL:
            # Left and right walls
            self.ball.position.x = g.BOARD_WALL if self.ball.position.x <= g.BOARD_WALL else g.BOARD_WIDTH - self.ball.size.x - g.BOARD_WALL
            self.ball.velocity.x *= -1
            self.collision_happened = True
        elif self.ball.position.y <= g.BOARD_WALL or self.ball.position.y + self.ball.size.y >= g.BOARD_HEIGHT - g.BOARD_WALL:
            # Top and bottom wall
            if self.ball.position.y <= g.BOARD_WALL:
                self.scores[g.ID_PLAYER1] += 1
            else:
                self.scores[g.ID_PLAYER2] += 1

            self.particle_pool.reset(self.ball.position.x + self.ball.size.x / 2, self.ball.position.y + self.ball.size.y / 2)
            self.who_serves = not self.who_serves
            self.score_happened = True
            self.reset_ball()

            if self.scores[g.ID_PLAYER1] >= g.POINTS_TO_WIN or self.scores[g.ID_PLAYER2] >= g.POINTS_TO_WIN:
                self.status = g.STATUS_ENDED
                self.victory_happened = True
                self.ball.velocity.x = 0
                self.ball.velocity.y = 0


    def update(self, dt):

        self.t += dt

        if self.t < 1 or self.status not in [g.STATUS_ACTIVE, g.STATUS_ENDED]:
            return

        self.collision_happened = False
        self.victory_happened = False
        self.score_happened = False

        self.update_paddle_position(self.player1, dt)
        self.update_paddle_position(self.player2, dt)

        if self.particle_pool.get_n_actives() > 0:
            self.particle_pool.update(dt)
            return

        self.update_ball_position(dt)
